Add new Datadog scaler

arapulido opened the pull request:
This PR adds a new Datadog scaler. Right now, it uses Datadog metrics to drive HPA, both on a target value and a target average value. In the future, it could grow to also allow to scale based on events, monitor statuses, etc.

Related docs PR: https://github.com/kedacore/keda-docs/pull/602

This is an example of what the TriggerAuthentication and the ScaledObject definitions would look like:

```yaml
apiVersion: v1
kind: Secret
metadata:
  name: datadog-secrets
  namespace: my-project
type: Opaque
data:
  apiKey: # Required: base64 encoded value of Datadog apiKey
  appKey: # Required: base64 encoded value of Datadog appKey
  datadogSite: # Optional: base64 encoded value of Datadog site
---
apiVersion: keda.sh/v1alpha1
kind: TriggerAuthentication
metadata:
  name: keda-trigger-auth-datadog-secret
  namespace: my-project
spec:
  secretTargetRef:
    # Required: API key for your Datadog account
  - parameter: apiKey
    name: datadog-secrets
    key: apiKey
    # Required: APP key for your Datadog account
  - parameter: appKey
    name: datadog-secrets
    key: appKey
    # Optional: Datadog site. Default: "datadoghq.com"
  - parameter: datadogSite
    name: datadog-secrets
    key: datadogSite
---
apiVersion: keda.sh/v1alpha1
kind: ScaledObject
metadata:
  name: datadog-scaledobject
  namespace: my-project
spec:
  scaleTargetRef:
    name: worker
  triggers:
  - type: datadog
    metadata:
      # Required: datadog metric query
      query: "sum:trace.redis.command.hits{env:none,service:redis}.as_count()"
      # Required: according to the number of query result, to scale the TargetRef
      queryValue: "7"
      # Optional: (global or average). Whether the target value is global or average per pod. Default: average
      type: "global"
      # Optional: The time window (in seconds) to retrieve metrics from Datadog. Default: 90
      age: "60"
    authenticationRef:
      name: keda-trigger-auth-datadog-secret
```

### Checklist

- [x] Commits are signed with Developer Certificate of Origin (DCO - [learn more](https://github.com/kedacore/keda/blob/main/CONTRIBUTING.md#developer-certificate-of-origin-signing-your-work))
- [x] Tests have been added
- [x] A PR is opened to update our Helm chart ([repo](https://github.com/kedacore/charts)) *(if applicable, ie. when deployment manifests are modified)*
- [x] A PR is opened to update the documentation on ([repo](https://github.com/kedacore/keda-docs)) *(if applicable)*
- [x] Changelog has been updated

Relates to #1002

tomkerkhove commented:
Thank you for your PR! Would you mind opening a PR for docs as well please?
https://github.com/kedacore/keda-docs

zroubalik commented:
> Thank you for your PR! Would you mind opening a PR for docs as well please? https://github.com/kedacore/keda-docs

And e2e tests please :) https://github.com/kedacore/keda/tree/main/tests

arapulido commented:
> Thank you for your PR! Would you mind opening a PR for docs as well please? https://github.com/kedacore/keda-docs

PR for docs: https://github.com/kedacore/keda-docs/pull/602

arapulido commented:
> > Thank you for your PR! Would you mind opening a PR for docs as well please? https://github.com/kedacore/keda-docs
> 
> And e2e tests please :) https://github.com/kedacore/keda/tree/main/tests

I will work on those. Thanks!

arapulido commented:
Thanks both for the early feedback. I will be off next week. I will work on the e2e tests once I get back. Thanks again!

tomkerkhove commented:
Thanks and have a good week off!

JorTurFer commented on pkg/scalers/datadog_scaler.go:
```diff
@@ -0,0 +1,316 @@
+		meta.vType = average // Default to average between pods
+	}
+
+	if val, ok := config.TriggerMetadata["metricName"]; ok {
+		meta.metricName = GenerateMetricNameWithIndex(config.ScalerIndex, kedautil.NormalizeString(fmt.Sprintf("datadog-%s", val)))
+	} else {
+		return nil, fmt.Errorf("no metric name given")
+	}
```
Does requiring `metricName` make sense? I mean, `metricName` it's only used to generate the metric name in the scaler, maybe it could be optional and in case of not be provided, generate it in base of the query for example.
There are several examples of this last behavior in other scalers that you could use.

zroubalik replied:
+1 

I think that we don't need to enforce the metricName (as it was just for KEDA internal usage). The current implementation ensures that the generated `metricName` is unique, so I think we can drop this one and make the spec simpler.  Unless there's a serious need for it.

arapulido replied:
ah, cool, let me try removing it. And no, there is no need

JorTurFer commented on pkg/scalers/datadog_scaler.go:
```diff
@@ -0,0 +1,316 @@
+		return nil, fmt.Errorf("no app key given")
+	}
+
+	if val, ok := config.AuthParams["datadogSite"]; ok {
+		if val != "" {
+			meta.datadogSite = val
+		} else {
+			meta.datadogSite = "datadoghq.com"
+		}
+	} else {
+		meta.datadogSite = "datadoghq.com"
+	}
```
IMHO, we should avoid the prefix `datadog`, the scaler itself is already for datadog so it can be supposed that all the metadata are related with datadog.
WDYT?

zroubalik replied:
I think that it is okay to keep the value as `datadogSite` as it is explanatory and it is not that long.

arapulido replied:
Yes, I would prefer keeping datadogSite, as just site is too ambiguous

JorTurFer reviewed (changes requested):
Thanks for your effort! ❤️ 
Apart from the e2e test, 2 small nits

arschles commented on pkg/scalers/datadog_scaler.go:
```diff
@@ -0,0 +1,316 @@
+		return nil, fmt.Errorf("no app key given")
+	}
+
+	if val, ok := config.AuthParams["datadogSite"]; ok {
+		if val != "" {
+			meta.datadogSite = val
+		} else {
+			meta.datadogSite = "datadoghq.com"
+		}
+	} else {
+		meta.datadogSite = "datadoghq.com"
+	}
```
this could be made a bit shorter and simpler with:

```suggestion
        siteVal = "datadoghq.com"
	if val, ok := config.AuthParams["datadogSite"]; ok && val != "" {
	        siteVal = val
	}
	meta.datadogSite = "datadoghq.com"
```

zroubalik replied:
+1

arschles commented on pkg/scalers/datadog_scaler.go:
```diff
@@ -0,0 +1,316 @@
+func newDatadogConnection(ctx context.Context, meta *datadogMetadata) (*datadog.APIClient, error) {
+
+	ctx = context.WithValue(
+		context.Background(),
```
why not use the given `ctx` parameter here rather than creating a new context root?

```suggestion
		ctx,
```

zroubalik replied:
+1

arschles commented on pkg/scalers/datadog_scaler.go:
```diff
@@ -0,0 +1,316 @@
+func (s *datadogScaler) IsActive(ctx context.Context) (bool, error) {
+
+	ctx = context.WithValue(
+		context.Background(),
```
similar to the above, why not use the given `ctx` parameter here?

```suggestion
		ctx,
```

arschles commented on pkg/scalers/datadog_scaler.go:
```diff
@@ -0,0 +1,316 @@
+func (s *datadogScaler) getQueryResult(ctx context.Context) (int, error) {
+
+	ctx = context.WithValue(
+		context.Background(),
```
```suggestion
		ctx,
```

arschles commented on pkg/scalers/datadog_scaler.go:
```diff
@@ -0,0 +1,316 @@
+
+	points := series[0].GetPointlist()
+
+	if len(points) == 0 {
+		return 0, fmt.Errorf("no Datadog metrics returned")
+	}
+
+	var val int = int(*points[0][1])
+
+	return val, nil
```
do we need to check whether `*points[0]` has 2 (exactly, or more than) elements in it? WDYT of something like the below?

```suggestion
	if len(points) == 0 || len(points[0]) <= 2 {
		return 0, fmt.Errorf("no Datadog metrics returned")
	}

	return int(*points[0][1]), nil
```

arschles commented on pkg/scalers/datadog_scaler.go:
```diff
@@ -0,0 +1,316 @@
+
+	targetQueryValue := resource.NewQuantity(int64(s.metadata.queryValue), resource.DecimalSI)
+
+	switch s.metadata.vType {
+
+	case average:
+		externalMetric = &v2beta2.ExternalMetricSource{
+			Metric: v2beta2.MetricIdentifier{
+				Name: s.metadata.metricName,
+			},
+			Target: v2beta2.MetricTarget{
+				Type:         v2beta2.AverageValueMetricType,
+				AverageValue: targetQueryValue,
+			},
+		}
+	case global:
+		externalMetric = &v2beta2.ExternalMetricSource{
+			Metric: v2beta2.MetricIdentifier{
+				Name: s.metadata.metricName,
+			},
+			Target: v2beta2.MetricTarget{
+				Type:  v2beta2.ValueMetricType,
+				Value: targetQueryValue,
+			},
+		}
+	}
```
I think we need a `default` case here so that `externalMetric` is not `nil`. otherwise, `externalMetric` needs to be allocated above. WDYT?

arschles commented on pkg/scalers/datadog_scaler.go:
```diff
@@ -0,0 +1,316 @@
+		Timestamp:  metav1.Now(),
+	}
+
+	return append([]external_metrics.ExternalMetricValue{}, metric), nil
```
```suggestion
	return []external_metrics.ExternalMetricValue{metric}, nil
```

arapulido replied:
mmm, I don't think that change is correct. GetMetrics needs to return a slice of external_metrics.ExternalMetricValue struct, which is what I am doing there. Am I missing anything?

arapulido replied:
@arschles ? ☝️

zroubalik replied:
I think it is ok to keep it this way :)

arschles replied:
sorry for the delay! I lost track of this PR.

the `append` is technically correct. I was just suggesting the change because it avoids the extra `append`.

arapulido commented:
Hello all! Thanks again for the reviews! I think I have addressed all comments and I have pushed the E2E tests. Let me know how we can do it to add the needed secrets to your pipeline.

JorTurFer commented:
> Hello all! Thanks again for the reviews! I think I have addressed all comments and I have pushed the E2E tests. Let me know how we can do it to add the needed secrets to your pipeline.

Hi @arapulido ,
Thanks for this!!!
I guess that we have to request any free access account for datadog, let me know what secrets are needed and I will request the account and add the secrets to the pipeline (sorry, I have 0 exprtise with datadog)

JorTurFer commented:
/run-e2e datadog.test.*

zroubalik reviewed (commented):
Generally looking good, I have 2 questions:
1. I can see that you are using `datadog-api-client-go/api/v1`, though it seems like there is `v2` as well. Is there a particular reason to not use `v2`?
2. Do you think you can use KEDA http client? (it allows setting a global timeout), seems like the datadog client might allow this: https://github.com/DataDog/datadog-api-client-go/blob/4cb420c458f3e0259afd11a0a918108ebd8cbd32/api/v1/datadog/client.go#L118-L121
Example usage of KEDA http client: https://github.com/kedacore/keda/blob/eeefb100ad9418a3add72b481d2495feb0318d99/pkg/scalers/metrics_api_scaler.go#L75

JorTurFer commented:
/run-e2e datadog.test.*

JorTurFer commented:
/run-e2e datadog.test.*

JorTurFer commented:
Hi @arapulido 
I have tried again the e2e tests after downgrade to free plan and they fail, I think that is not related with Datadog plan but maybe you could take a look.
Thanks in advance ❤️

arapulido commented:
/run-e2e datadog.test.*

arapulido commented:
> Hi @arapulido I have tried again the e2e tests after downgrade to free plan and they fail, I think that is not related with Datadog plan but maybe you could take a look. Thanks in advance ❤️

I cannot reproduce this failure in my environment. It is failing creating a secret, but not sure why. What would be the best way to debug this in the CI environment, @JorTurFer ?

JorTurFer commented:
/run-e2e datadog.test.*

JorTurFer commented on tests/scalers/datadog.test.ts:
```diff
@@ -0,0 +1,286 @@
+      --namespace ${datadogNamespace} -o jsonpath="{.status.numberReady}"`).stdout
+  }
+
+  const tmpFile = tmp.fileSync()
+  fs.writeFileSync(tmpFile.name, deployYaml)
+  sh.exec(`kubectl create namespace ${testNamespace}`)
+  t.is(
+    0,
+    sh.exec(`kubectl apply -f ${tmpFile.name} --namespace ${testNamespace}`).code,
+    'creating a deployment should work.'
+  )
+
+  t.is(
+    0,
+    sh.exec(`kubectl create secret generic datadog-secrets --from-literal=apiKey=${datadogApiKey} \
+      --from-literal=appKey=${datadogAppKey} --from-literal=datadogSite=${datadogSite} --namespace ${testNamespace}`).code,
+    'creating a generic secret should work.'
+  )
+
```
Please, create first the secret and then the scaled object, there is a known bug we are working on that produces errors if the secrets are not available during the SO creation. In this case the SO is created before the secret and that why sometimes it works and other times not

```suggestion
  t.is(
    0,
    sh.exec(`kubectl create secret generic datadog-secrets --from-literal=apiKey=${datadogApiKey} \
      --from-literal=appKey=${datadogAppKey} --from-literal=datadogSite=${datadogSite} --namespace ${testNamespace}`).code,
    'creating a generic secret should work.'
  )

  const tmpFile = tmp.fileSync()
  fs.writeFileSync(tmpFile.name, deployYaml)
  sh.exec(`kubectl create namespace ${testNamespace}`)
  t.is(
    0,
    sh.exec(`kubectl apply -f ${tmpFile.name} --namespace ${testNamespace}`).code,
    'creating a deployment should work.'
  )

```

arapulido replied:
I have now changed the order as you suggested, and I also took the opportunity to improve the output messages in case of test failure, so we can understand it better.

JorTurFer commented:
> > Hi @arapulido I have tried again the e2e tests after downgrade to free plan and they fail, I think that is not related with Datadog plan but maybe you could take a look. Thanks in advance ❤️
> 
> I cannot reproduce this failure in my environment. It is failing creating a secret, but not sure why. What would be the best way to debug this in the CI environment, @JorTurFer ?
> 
> The line that fails the only thing it does is the following:
> 
> ```
> kubectl create secret generic datadog-secrets --from-literal=apiKey=${datadogApiKey} \
>       --from-literal=appKey=${datadogAppKey} --from-literal=datadogSite=${datadogSite} --namespace ${testNamespace}
> ```
> 
> can you make sure that the env variables that populate the API and APP keys are not empty? Thanks!

The second one is noise because the namespace was terminating and that was why it failed. I think that the problem could be in the creation order, before the SO and then the secret.

arapulido commented:
> Generally looking good, I have 2 questions:
> 
>     1. I can see that you are using `datadog-api-client-go/api/v1`, though it seems like there is `v2` as well. Is there a particular reason to not use `v2`?

I use an endpoint (QueryMetrics) that is not yet available in v2.

> 
>     2. Do you think you can use KEDA http client? (it allows setting a global timeout), seems like the datadog client might allow this: https://github.com/DataDog/datadog-api-client-go/blob/4cb420c458f3e0259afd11a0a918108ebd8cbd32/api/v1/datadog/client.go#L118-L121
>        Example usage of KEDA http client: https://github.com/kedacore/keda/blob/eeefb100ad9418a3add72b481d2495feb0318d99/pkg/scalers/metrics_api_scaler.go#L75

Sure, I have now pushed a commit with this. Thanks!

JorTurFer commented:
/run-e2e datadog.test.*

JorTurFer commented:
/run-e2e datadog.test.*

JorTurFer commented:
/run-e2e datadog.test.*

JorTurFer reviewed (commented):
LGTM!
Thanks a ton for the effort that you have done doing this PR ❤️ ❤️ ❤️ 
There is a conflic in CHANGELOG.md, could you solve it?

arapulido commented:
> LGTM! Thanks a ton for the effort that you have done doing this PR ❤️ ❤️ ❤️ There is a conflict in CHANGELOG.md, could you solve it?

Done! I have rebased with main and resolved the conflict. Thanks a lot for the reviews and the help!

JorTurFer commented:
/run-e2e datadog.test.*

arschles commented on pkg/scalers/datadog_scaler.go:
```diff
@@ -0,0 +1,307 @@
+
+// GetMetricSpecForScaling returns the MetricSpec for the Horizontal Pod Autoscaler
+func (s *datadogScaler) GetMetricSpecForScaling(context.Context) []v2beta2.MetricSpec {
+	var externalMetric *v2beta2.ExternalMetricSource
+	externalMetric = new(v2beta2.ExternalMetricSource)
```
```suggestion
	externalMetric := new(v2beta2.ExternalMetricSource)
```

arschles commented on pkg/scalers/datadog_scaler.go:
```diff
@@ -0,0 +1,307 @@
+				Type:  v2beta2.ValueMetricType,
+				Value: targetQueryValue,
+			},
+		}
```
do we need to fail here if there is no value value for `s.metadata.vType`? If, for some reason, it's not `average` or `global`, `externalMetric` will be `nil`

arapulido replied:
the type is defaulted to Average, so I it won't be empty. In any case, you mentioned this in your previous review, and that's why I added the "new" constructor, as in that case, the pointer won't be nil: https://go.dev/play/p/ROR4tfKKQl0

zroubalik reviewed (approved):
LGTM! Great job :)

bmpandrade commented:
So good to see this being merged... thanks to all who contributed to make this happen 🙌 

Is there a possibility to release this sooner, or v2.6.0 is only due to feb 24 as the milestone states?

JorTurFer commented:
Hi @bmpandrade 
You could use `main` tag if you'd like to try this scaler before the release.

zroubalik commented:
@bmpandrade we might actually do this release a little bit sooner, maybe end of this month. Be careful, consuming KEDA from `main` is obviously not stable and not for production :)

bmpandrade commented:
Thank you both :) 

At this point, would be mostly for experimentation so we would be fine with ref `main` for now until we have a release out.

arschles commented:
@bmpandrade just an idea: for now, until the release, you could pull `main` down, tag it, and push it to your own image repository. that would ensure that you only have this change and not anything more that gets merged between now and the release.
